Use RAII stream guard in Operator

HuitongQiu opened the pull request:
Summary: Currently we call `SwitchToDevice` to run operations on a device, which the device and stream are not automatically restored afterwards. We want to use a RAII guard, `StreamGuard`, to do the switching and automatically restore the device and stream afterwards.

Differential Revision: D14529276

ezyang commented on aten/src/ATen/core/context_base.h:
```diff
@@ -44,6 +45,10 @@ class CAFFE2_API BaseContext {
 
+  virtual Stream GetStream(int /*stream_id*/) {
+    return Stream{Stream::DEFAULT, Device(DeviceType::CPU, -1)};
+  }
```
I don't think you should default this method. Fine for testing but we should follow `SwitchToDevice`.

HuitongQiu replied:
I'll make it a pure virtual function.

ezyang reviewed (approved):
Code change looks good, but with the comment about making it pure virtual. The crux of the matter will be in the testing!

HuitongQiu commented:
@pytorchbot retest this please

HuitongQiu commented:
@pytorchbot retest this please

HuitongQiu commented:
@pytorchbot retest this please

HuitongQiu commented:
@pytorchbot retest this please

HuitongQiu commented:
@pytorchbot retest this please

HuitongQiu commented:
@pytorchbot retest this please

Commit:
Use RAII stream guard in Operator (#18222)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18222

Currently we call `SwitchToDevice` to run operations on a device, which the device and stream are not automatically restored afterwards. We want to use a RAII guard, `StreamGuard`, to do the switching and automatically restore the device and stream afterwards.

Reviewed By: ezyang

Differential Revision: D14529276

fbshipit-source-id: e613be3650395e199c56dd66a0a0d3d4c2cdcd43
